### ISSUE TITLE
rclone Keeps Failing to Copy File with "corrupted on transfer" (local->crypt->mount)

### ISSUE THREAD

**Issue by irkcow** [rclone/rclone#4031]
#### What is the problem you are having with rclone?
Than I crypt local folder and mount it on windows 10, i get this massage:
[log.txt](https://github.com/rclone/rclone/files/4279999/log.txt)

"ERROR : IO error: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "cd60ea4dd2e9e81e702bc637df404751" vs "35f640a1fd3ae61ad81581496f86a60b"

#### What is your rclone version (output from `rclone version`)
rclone v1.51.0
- os/arch: windows/amd64
- go version: go1.13.7

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Windows 10 64bit

####  Which cloud storage system are you using? (eg Google Drive)
Local

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
.\rclone mount lo_crpt: p:  --vfs-cache-mode writes

[lo_crpt]
type = crypt
remote = e:\\iokpb_cr\\
filename_encryption = off
directory_name_encryption = false
password = *** ENCRYPTED ***
password2 = *** ENCRYPTED ***

Then i make text file in note and save and get error,
without key --vfs-cache-mode the system is work, but i can not edit the file

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

The service rclone has been started.
2020/03/03 16:04:49 ERROR : kjdfhgjdfkdfk.txt: Failed to copy: corrupted on transfer: MD5 crypted hash differ "7cfd006a83f095b44dc1a4ea0ed75da4" vs "b077f73d494a4d595a490dedc46ae797"
2020/03/03 16:04:49 ERROR : kjdfhgjdfkdfk.txt(0xc000112480): failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "7cfd006a83f095b44dc1a4ea0ed75da4" vs "b077f73d494a4d595a490dedc46ae797"
2020/03/03 16:04:49 ERROR : kjdfhgjdfkdfk.txt(0xc000112480): RWFileHandle.Flush error: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "7cfd006a83f095b44dc1a4ea0ed75da4" vs "b077f73d494a4d595a490dedc46ae797"
2020/03/03 16:04:49 ERROR : IO error: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "7cfd006a83f095b44dc1a4ea0ed75da4" vs "b077f73d494a4d595a490dedc46ae797"

**Comment by ncw** [comment]
> Then i make text file in note and save and get error,

Which editor are you using to make the text file?

Have you tried a different one?

Can you attach a log with `-vv` of the problem please?

I suspect that somehow the file is still open when we try to transfer it but the debug log should show the problem.

**Comment by irkcow** [comment]
I attached the log file in first message, editor is notepad. But editor is
not important thing. Then I copy the file to mounted disk, the situation is
the same


вт, 3 мар. 2020 г., 20:36 Nick Craig-Wood <notifications@github.com>:

**Comment by ncw** [comment]
> I attached the log file in first message

Sorry, I missed that

> editor is notepad. But editor is not important thing. Then I copy the file to mounted disk, the situation is the same

OK.

I will try to reproduce.

**Comment by nickh13** [comment]
I have the same problem (all the same, but Windows7/2008R2 x64).
I try 1.51, 1.50 ,1.49 - it's not work.
But then I try 1.48 - it's working!
I try use FAR Manager or Windows Explorer, it does not matter.
Winfsp 2019.1... 

... But then I reinstall WinFsp to latest v.16 (2020) everything started working fine with rclone 1.51 !

Thanks for your issue, while I wrote this answer, while digging through the versions, I finally figured out the problem :)
I hope you can do it too!

**Comment by ncw** [comment]
Hmm... Have you got a recipe for reproduction?

**Comment by nickh13** [comment]
Windows 7/2008R2 x64
RClone 1.51
Winfsp v.16 (2020)

[LocalEncryptedBackup]
type = crypt
remote = E:\LocalEncryptedBackup
filename_encryption = standard
directory_name_encryption = false
password = ***************
password2 = ***************

C:\tools\rclone.exe --vfs-cache-mode=writes mount LocalEncryptedBackup: z:

**Comment by nickh13** [comment]
When I mount a drive without caching, when I try to copy to a mounted drive using FAR, the system asks for administrative rights, although the mount itself was started without them.
In this case, these errors come out:

The service rclone has been started.
2020/04/01 00:43:36 ERROR : 123/track_1.mkv: WriteFileHandle: ReadAt: Can't read and write to file without --vfs-cache-mode >= minimal
2020/04/01 00:43:36 ERROR : 123/track_1.mkv: WriteFileHandle: Truncate: Can't change size without --vfs-cache-mode >= writes
2020/04/01 00:43:36 ERROR : 123/track_1.mkv: WriteFileHandle: Can't open for write without O_TRUNC on existing file without --vfs-cache-mode >= writes
2020/04/01 00:43:36 ERROR : 123/track_1.mkv: WriteFileHandle: Can't open for write without O_TRUNC on existing file without --vfs-cache-mode >= writes

After obtaining admin rights, the process of copying without mounting is successful ...
But for each next file this error is issued here:
2020/04/01 00:44:32 ERROR : 123/track_1.mkv: WriteFileHandle: Truncate: Can't change size without --vfs-cache-mode >= writes
2020/04/01 00:46:27 ERROR : 123/track_2.mkv: WriteFileHandle: Truncate: Can't change size without --vfs-cache-mode >= writes

**Comment by nickh13** [comment]
And when I try to mount with the parameter  --vfs-cache-mode=minimal 
That such errors are issued here:
The service rclone has been started.
2020/04/01 01:08:09 ERROR : 123/1.zip: Failed to copy: corrupted on transfer: MD5 crypted hash differ "29d6cb7890bbb8328f0d2a7c596981a7" vs "8239aae4e224cdc677a6c6fb04924b21"
2020/04/01 01:08:09 ERROR : 123/1.zip(0xc00006c900): failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "29d6cb7890bbb8328f0d2a7c596981a7" vs "8239aae4e224cdc677a6c6fb04924b21"
2020/04/01 01:08:09 ERROR : 123/1.zip(0xc00006c900): RWFileHandle.Release error: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "29d6cb7890bbb8328f0d2a7c596981a7" vs "8239aae4e224cdc677a6c
6fb04924b21"
2020/04/01 01:08:09 ERROR : IO error: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "29d6cb7890bbb8328f0d2a7c596981a7" vs "8239aae4e224cdc677a6c6fb04924b21"

**Comment by nickh13** [comment]
RClone 1.48 works fine with caching (no errors)

Errors appear without caching, but everything works.

The service rclone_1.48.0 has been started.
2020/04/01 01:15:21 ERROR : 123/1.zip: WriteFileHandle: Truncate: Can't change size without --vfs-cache-mode >= writes
2020/04/01 01:15:37 ERROR : 123/track_1.mkv: WriteFileHandle: Truncate: Can't change size without --vfs-cache-mode >= writes

At different points in time, I would like to work with and without caching (at example big files writes).

**Comment by snaix** [comment]
I Have the same problem,:
rclone version:
rclone v1.52.2
- os/arch: windows/amd64
- go version: go1.14.4

winfsp-1.7.20172
Windows 10 Pro 1909

**Comment by ncw** [comment]
Can you try [the latest beta](https://beta.rclone.org) please?

**Comment by irkcow** [comment]
**All the same**

2020/08/01 13:12:11 ERROR : IO error: vfs reader: failed to write to cache file: The system cannot find the file specified.
2020/08/01 13:12:30 ERROR : hello.txt: Failed to copy: corrupted on transfer: MD5 crypted hash differ "67391aeac5fc5996d8c098f15905c6a5" vs "490d3d3c5080e7986f4b98953d0eee29"
2020/08/01 13:12:30 ERROR : hello.txt: vfs cache: failed to upload try #5, will retry in 2m40s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "67391aeac5fc5996d8c098f15905c6a5" vs "490d3d3c5080e7986f4b98953d0eee29"
2020/08/01 13:15:10 ERROR : hello.txt: Failed to copy: corrupted on transfer: MD5 crypted hash differ "ec0a44905cace37694699ee8e15be546" vs "490d3d3c5080e7986f4b98953d0eee29"
2020/08/01 13:15:10 ERROR : hello.txt: vfs cache: failed to upload try #6, will retry in 5m0s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "ec0a44905cace37694699ee8e15be546" vs "490d3d3c5080e7986f4b98953d0eee29"
2020/08/01 13:20:10 ERROR : hello.txt: Failed to copy: corrupted on transfer: MD5 crypted hash differ "fa82120eec2e9836e5be3bd78629cfb0" vs "490d3d3c5080e7986f4b98953d0eee29"
2020/08/01 13:20:10 ERROR : hello.txt: vfs cache: failed to upload try #7, will retry in 5m0s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "fa82120eec2e9836e5be3bd78629cfb0" vs "490d3d3c5080e7986f4b98953d0eee29"

**Comment by ncw** [comment]
@snaix can you post your config with the secrets removed and the rclone command you used on the mount.

Also if you could post a sequence of steps for me to reproduce then I'll see if I can reproduce it here - thanks!

**Comment by nickh13** [comment]
I try latest beta
rclone v1.52.2-281-g324077fb-beta, - os/arch: windows/amd64, - go version: go1.14.6
WinFsp 2019.3

[LocalEncryptedBackup]
type = crypt
remote = D:\RCloneTest
filename_encryption = standard
directory_name_encryption = false
...

It's works:
- without cache:   like early rclone_1.48.0 with error on all written files "ERROR : 1.zip: WriteFileHandle: Truncate: Can't change size without --vfs-cache-mode >= writes"
- with  --vfs-cache-mode=writes: All OK, no errors or warnings, and the files are written successfully

But several times I got a strange result that the file was 2x larger and the first part of its content was filled with 0.
I have not yet found how this happens.

I will test more further

**Comment by irkcow** [comment]
windows 10 pro x64
winfsp 2020 1.7.20172


[lo_crpt]
type = crypt
remote = d:\\\\iokpb_cr\\\\
filename_encryption = off
directory_name_encryption = false
password = *** ENCRYPTED ***
password2 = *** ENCRYPTED ***

PS K:\portable\rclone> .\rclone version
rclone v1.52.2-281-g324077fb-beta
- os/arch: windows/amd64
- go version: go1.14.6


PS K:\portable\rclone> .\rclone mount lo_crpt: p:  --vfs-cache-mode writes
Enter configuration password:
password:
The service rclone has been started.
2020/08/04 17:33:23 ERROR : 776543.txt: Failed to copy: corrupted on transfer: MD5 crypted hash differ "fa509e001397d0ba380022fa8573d4a7" vs "7c2dc5d372bdae9a7d9833a9a1262cff"
2020/08/04 17:33:23 ERROR : 776543.txt: vfs cache: failed to upload try #1, will retry in 10s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "fa509e001397d0ba380022fa8573d4a7" vs "7c2dc5d372bdae9a7d9833a9a1262cff"
2020/08/04 17:33:33 ERROR : 776543.txt: Failed to copy: corrupted on transfer: MD5 crypted hash differ "9826f255b17799ec3747d2e165ce3938" vs "7c2dc5d372bdae9a7d9833a9a1262cff"
2020/08/04 17:33:33 ERROR : 776543.txt: vfs cache: failed to upload try #2, will retry in 20s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: MD5 crypted hash differ "9826f255b17799ec3747d2e165ce3938" vs "7c2dc5d372bdae9a7d9833a9a1262cff"

**Comment by s0ftice** [comment]
Had the same issue with `MD5 crypted hash differ` when using (any) vfs-cache-mode on macOS 10.15.6 and rclone v1.52.2.
I have a crypt mount mapping to a local directory.

No issue at all with the latest rclone v1.52.3-290-g82975109-beta

**Comment by snaix** [comment]
@ncw 
Hi, sorry for late, my config as following:

[remote]
type = s3
provider = Other
env_auth = false
access_key_id =  *** ENCRYPTED ***
secret_access_key =  *** ENCRYPTED ***
region =  *** ENCRYPTED ***
endpoint = *** ENCRYPTED ***
acl = private

The warning message was shown after file transferring, and the file will be missing.
In my test, a big file(about 1GB) tranferring are more prone to this problem, seems a chunk error?

the mount command and 2 errors:

c:\Bin\rclone>rclone mount --cache-dir X:\rclone-vfs-cache\  --vfs-cache-max-age 0h5m0s --vfs-cache-mode=writes remote:kenc-cg S:
The service rclone has been started.
2020/08/19 15:08:07 ERROR : videoplayback_hdr.webm: corrupted on transfer: MD5 hash differ "6f31dc0f46c608613c45415603fa3b2b" vs "356884ff00ea0d55531364cf110a0391"
2020/08/19 15:08:07 ERROR : videoplayback_hdr.webm(0xc0000d3900): failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "6f31dc0f46c608613c45415603fa3b2b" vs "356884ff00ea0d55531364cf110a0391"
2020/08/19 15:08:07 ERROR : videoplayback_hdr.webm(0xc0000d3900): RWFileHandle.Flush error: failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "6f31dc0f46c608613c45415603fa3b2b" vs "356884ff00ea0d55531364cf110a0391"
2020/08/19 15:08:07 ERROR : IO error: failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "6f31dc0f46c608613c45415603fa3b2b" vs "356884ff00ea0d55531364cf110a0391"
2020/08/19 15:27:35 ERROR : videoplayback.webm: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/19 15:27:36 ERROR : videoplayback.webm(0xc000584040): failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/19 15:27:36 ERROR : videoplayback.webm(0xc000584040): RWFileHandle.Flush error: failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/19 15:27:36 ERROR : IO error: failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/19 15:29:18 ERROR : videoplayback.webm: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/19 15:29:18 ERROR : videoplayback.webm(0xc000584040): failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/19 15:29:18 ERROR : videoplayback.webm(0xc000584040): RWFileHandle.Release error: failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/19 15:29:18 ERROR : IO error: failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"

**Comment by snaix** [comment]
And I will test a beta version for this case.

**Comment by ncw** [comment]
I'm pretty sure the beta will fix this as you are using --vfs-cache-mode=writes  :crossed_fingers:

**Comment by snaix** [comment]
Hi @ncw 

Unfortunately, I still have this problem.

The beta rclone version is: 
rclone v1.52.3-328-g0f7a2f0f-beta
- os/arch: windows/amd64
- go version: go1.15

I run the command as:
rclone mount --cache-dir X:\rclone-vfs-cache\  --vfs-cache-mode=writes remote:kenc-cg S:

And copy a file about 1GB to driver S. The file is quickly copied to the cache, and the upload begins.

After uploading(the NIC traffic return to zero), I get the error:
2020/08/25 23:02:30 ERROR : videoplayback.webm: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/25 23:02:30 ERROR : videoplayback.webm: vfs cache: failed to upload try #1, will retry in 10s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/25 23:05:38 ERROR : videoplayback.webm: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"
2020/08/25 23:05:38 ERROR : videoplayback.webm: vfs cache: failed to upload try #2, will retry in 20s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: MD5 hash differ "0082dc98c2d1f374cec6f0235ded7d18" vs "362477d565ae74b6c5619dfe719c78cc"

The MD5 "0082dc98c2d1f374cec6f0235ded7d18" is the original file, I don't know what "362477d565ae74b6c5619dfe719c78cc" means.
The problem can be reproduced
BTW, I use a S3 compatible storage but not the amazon's.

I will trace the code for this problem, if I have time. :)
T

**Comment by brettferdosi** [comment]
In case this helps anyone, I can reproduce the problem reliably with the following:
```
rclone v1.57.0
- os/version: darwin 11.5.2 (64 bit)
- os/kernel: 20.6.0 (x86_64)
- os/type: darwin
- os/arch: amd64
- go/version: go1.16.6
- go/linking: dynamic
- go/tags: cmount
```
The config is
```
[test-crypt]
type = crypt
remote = /tmp/testlocal
```
And command
`rclone mount --vfs-cache-mode writes test-crypt: /tmp/testlocaldecrypted -v`


If you go into the mounted dir, initialize a git repository, then create and commit two text files in two separate commits using your editor to write the commit messages (mine is vim), you'll see output from rclone like:

```
2021/11/19 11:54:25 ERROR : gitdir/.git/COMMIT_EDITMSG: Failed to copy: corrupted on transfer: md5 crypted hash differ "39896530df8aea84938e9af9eaaccbf6" vs "19f7f2c141b31ee3df90e012a321dbc1"
2021/11/19 11:54:25 ERROR : gitdir/.git/COMMIT_EDITMSG: vfs cache: failed to upload try #1, will retry in 10s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: md5 crypted hash differ "39896530df8aea84938e9af9eaaccbf6" vs "19f7f2c141b31ee3df90e012a321dbc1"
2021/11/19 11:54:25 INFO  : gitdir/.git/objects/23/3c7c5f3ca77c67facf1fa64255ea183095999f: Copied (new)
2021/11/19 11:54:25 INFO  : gitdir/.git/objects/23/3c7c5f3ca77c67facf1fa64255ea183095999f: vfs cache: upload succeeded try #1
2021/11/19 11:54:25 ERROR : gitdir/.git/logs/HEAD: Failed to copy: corrupted on transfer: md5 crypted hash differ "602c3b86e058ccb710d2717f52b3171f" vs "56cdd7f8709e6ec2a19220b634ff0ccf"
2021/11/19 11:54:25 ERROR : gitdir/.git/logs/HEAD: vfs cache: failed to upload try #1, will retry in 10s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: md5 crypted hash differ "602c3b86e058ccb710d2717f52b3171f" vs "56cdd7f8709e6ec2a19220b634ff0ccf"
2021/11/19 11:54:25 ERROR : gitdir/.git/logs/refs/heads/master: Failed to copy: corrupted on transfer: md5 crypted hash differ "072eed828f3317ba4187dd95f6e53ec9" vs "33e3e792d1956e39c687a0223976d2b4"
2021/11/19 11:54:25 ERROR : gitdir/.git/logs/refs/heads/master: vfs cache: failed to upload try #1, will retry in 10s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: md5 crypted hash differ "072eed828f3317ba4187dd95f6e53ec9" vs "33e3e792d1956e39c687a0223976d2b4"
2021/11/19 11:54:25 INFO  : gitdir/.git/refs/heads/master: Copied (new)
2021/11/19 11:54:25 INFO  : gitdir/.git/refs/heads/master: vfs cache: upload succeeded try #1
2021/11/19 11:54:25 INFO  : gitdir/.git/MERGE_RR: Copied (new)
2021/11/19 11:54:25 INFO  : gitdir/.git/MERGE_RR: vfs cache: upload succeeded try #1
2021/11/19 11:54:34 INFO  : vfs cache: cleaned: objects 33 (was 33) in use 3, to upload 3, uploading 0, total size 25.233Ki (was 25.233Ki)
2021/11/19 11:54:35 ERROR : gitdir/.git/COMMIT_EDITMSG: Failed to copy: corrupted on transfer: md5 crypted hash differ "7dde140d4e1e811a3eae3435f9f421f0" vs "19f7f2c141b31ee3df90e012a321dbc1"
2021/11/19 11:54:35 ERROR : gitdir/.git/COMMIT_EDITMSG: vfs cache: failed to upload try #2, will retry in 20s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: md5 crypted hash differ "7dde140d4e1e811a3eae3435f9f421f0" vs "19f7f2c141b31ee3df90e012a321dbc1"
2021/11/19 11:54:35 ERROR : gitdir/.git/logs/refs/heads/master: Failed to copy: corrupted on transfer: md5 crypted hash differ "ff730bd8cfc2f365eccbdbe4a638e11f" vs "33e3e792d1956e39c687a0223976d2b4"
2021/11/19 11:54:35 ERROR : gitdir/.git/logs/HEAD: Failed to copy: corrupted on transfer: md5 crypted hash differ "9ffbf357a28bc5bbff321152846a839b" vs "56cdd7f8709e6ec2a19220b634ff0ccf"
2021/11/19 11:54:35 ERROR : gitdir/.git/logs/refs/heads/master: vfs cache: failed to upload try #2, will retry in 20s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: md5 crypted hash differ "ff730bd8cfc2f365eccbdbe4a638e11f" vs "33e3e792d1956e39c687a0223976d2b4"
2021/11/19 11:54:35 ERROR : gitdir/.git/logs/HEAD: vfs cache: failed to upload try #2, will retry in 20s: vfs cache: failed to transfer file from cache to remote: corrupted on transfer: md5 crypted hash differ "9ffbf357a28bc5bbff321152846a839b" vs "56cdd7f8709e6ec2a19220b634ff0ccf"
2021/11/19 11:54:55 ERROR : gitdir/.git/COMMIT_EDITMSG: Failed to copy: corrupted on transfer: md5 crypted hash differ "a1cf63b489dfd89518b8f9f1e92b9bc0" vs "19f7f2c141b31ee3df90e012a321dbc1"
2021/11/19 11:54:55 ERROR : gitdir/.git/logs/HEAD: Failed to copy: corrupted on transfer: md5 crypted hash differ "404e2b569b04846ead0529a655c8f800" vs "56cdd7f8709e6ec2a19220b634ff0ccf"
2021/11/19 11:54:55 ERROR : gitdir/.git/logs/refs/heads/master: Failed to copy: corrupted on transfer: md5 crypted hash differ "f630d313afe18d3bae325d667f5259e0" vs "33e3e792d1956e39c687a0223976d2b4"
```

The files that fail to upload then don't ever make it out of the vfs cache. Note that across upload attempts the dst hashes remain constant and the src ones change. I think this may have something to do with files getting rapidly deleted/re-added/modified. Other files (like the text files committed in this example) make it out of the vfs cache fine. I saw similar behavior building a TeX document in the test mount, where certain build files, including the output pdf, got stuck.

You can work around the issue by adding `--no-checksum` to the rclone invocation. @ncw, is this a horrible idea?

**Comment by ncw** [comment]
I managed to replicate this with

```
mkdir /tmp/crypted
rclone mount -vv --vfs-cache-mode writes :crypt,remote=/tmp/crypted,password=$(rclone obscure password): /mnt/tmp 2>&1 | tee log.crypted
```

and

```
cd /mnt/tmp/
git init z
cd z
date -Is > file ; git add file ; git commit -m commit
date -Is > file ; git add file ; git commit -m commit
```

This fails as above with

> ERROR : z/.git/logs/HEAD: Failed to copy: corrupted on transfer: md5 crypted hash differ "b1aeb852d822f4f672b07b70edabafa4" vs "e3cb513354f0ffaf84c2671ab832b877"

Note that that md5 crypted hash error is coming from the crypt backend

https://github.com/rclone/rclone/blob/d252816706a3226dbc055616df91f2e1793a0459/backend/crypt/crypt.go#L446

The hashes should both be different as we change the nonce on upload, so I think the hash of the destination local object must have been cached incorrectly

Commenting out the hash caching in the local backend fixes the problem so that is definitely on the right track...

Will carry on investigating tomorrow.

**Comment by ncw** [comment]
OK I think I've fixed this now.

It turned out to be the second hard problem in computer science "cache invalidation". (The first is "naming things" just in case you were wondering!).

Can you try this please?

[v1.58.0-beta.5910.b91c349cd.fix-4031-local-hash-cache](https://beta.rclone.org/branch/fix-4031-local-hash-cache/v1.58.0-beta.5910.b91c349cd.fix-4031-local-hash-cache/) on branch [fix-4031-local-hash-cache](https://github.com/rclone/rclone/tree/fix-4031-local-hash-cache) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.58

**Comment by KrLx1994roller** [comment]
Hi @ncw,

I'm just commenting here to let you know that latest beta fixes this issue :)
```
rclone v1.58.0-beta.5995.9cc50a614
- os/version: debian 10.11 (64 bit)
- os/kernel: 4.19.0-18-amd64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.17.6
- go/linking: static
- go/tags: none
```

I was getting the exact same crypt md5 mismatch error with NextCloud when renaming a file. I have an encrypted network directory mounted with fstab and decrypted using rclone. This is part of the service I use to mount the crypt using rclone:
```
/usr/bin/rclone mount hdd: /var/www/nextcloud/data/user/files \
        --config=/var/www/nextcloud/config/rclone.conf \
        --log-level INFO \
        --log-file /var/log/nextcloud/mounts/vol1.log \
        --vfs-cache-mode full \
        --vfs-cache-max-size 36G \
        --cache-dir /var/www/nextcloud/cache/vol1 \
        --allow-other
```

The latest beta fixes the issue. Thanks a lot for your incredible work! I'll show my appreciation with a couple of coffees for you sometime soon ;)
